docs/DesignDocs/extra_enclave_data.md

```diff
@@ -0,0 +1,142 @@
+Extra SGX enclave data
+======================
+
+This document proposes a mechanism for injecting "extra enclave data" into the
+SGX enclave memory image. The extra data is signed and measured as part of the
+enclave memory image and is therefore covered by MRENCLAVE.
+
+## Motivation
+
+This mechanism may be used by applications with memory requirements that cannot
+be met by the Open Enclave read-write heap. Examples include:
+
+- Read-executable pages intended for loaded program segments.
+- Additional guard pages used to protect user-defined stacks.
+- Read-only pages intended for non-writable program data.
+- Inclusion of trusted hashes.
+- Inclusion of trusted public keys.
+- Read-write-executable pages for user-defined program loaders.
+
+This feature allows an application to manage a contiguous region or memory,
+controlling its content and the permissions of each of its pages.
+
+## User Experience
+
+Users may add additional pages by defining a "hook" that is called by the
+OE SGX enclave loader. This hook calls a special function for adding pages.
+For example, an Open Enclave host application may define the following function
+to inject one read-write-executable page.
+
+```
+oe_result_t oe_load_extra_enclave_data_hook(void* arg, uint64_t baseaddr)
+{
+    __attribute__((__aligned__(PAGE_SIZE)))
+    uint8_t page[PAGE_SIZE];
+    uint64_t flags = 0;
+
+    memset(page, 0, sizeof(page));
+    flags |= SGX_SECINFO_REG;
+    flags |= SGX_SECINFO_R;
+    flags |= SGX_SECINFO_W;
+    flags |= SGX_SECINFO_X;
+    const bool extend = true;
+    oe_load_extra_enclave_data(arg, 0, page, flags, extend);
+
+    return OE_OK;
+}
+```
+
+This function calls **oe_load_extra_enclave_data()**, which adds the page to
```

Comment:
Would oe_load_extra_enclave_data be SGX-only or TEE-agnostic?  If SGX only, it's again the wrong name.